### ISSUE TITLE
Firefox Snap / KeepassXC flatpak: KeePassXC-Browser - Failed to connect: Unknown error

### ISSUE THREAD

**Issue by marcelklehr** [keepassxreboot/keepassxc-browser#1875]
## Expected Behavior

Connect my Firefox KeePassXC-Brower WebExtension to my KeepassXC flatpak

## Current Behavior

It does not.

## Possible Solution

I don't know if firefox snap on ubuntu 22.04 is currently supported or not, so that may be the issue, but in that case I recommend making that more clear as firefox snap is the default on ubuntu 22.04 afaik. 

## Steps to Reproduce (for bugs)

1. Start KeepassXC
2. start firefox snap
3. install KeepassXC-Browser
4. Enable Browser integration in KeepassXC
5. Open Extension in browser
6. See "Cannot connect to KeePassXC. Check that browser integration is enabled in KeePassXC settings."
7. Click Reload in Extension
8. KeepassXC window comes to front with unlocked DB, but nothing else
9. Go back to browser extension
10. See "Key exchange was not successful."

## Debug info

KeePassXC - Version 2.7.4
Revision: 63b2394
Distribution: Flatpak

Qt 5.15.8
Debugging mode is disabled.

Operating system: KDE Flatpak runtime
CPU architecture: x86_64
Kernel: linux 5.19.0-32-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 2.19.1

KeePassXC-Browser - 1.8.5.1
Operating system: Linux x86_64
Browser: Mozilla Firefox Snap For Ubuntu 110.0.1 (64-bit)

Operating system:

Ubuntu 22.04.2 LTS
GNOME 42.5
X11

### Troubleshooting results

* native_messaging-hosts file:

```
{
    "allowed_extensions": [
        "keepassxc-browser@keepassxc.org"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "/var/lib/flatpak/exports/bin/org.keepassxc.KeePassXC",
    "type": "stdio"
}
```

* keepassxc-proxy is not running
* `sudo strace -f -p $(pgrep firefox) 2>&1 | grep keepass` yields no `openat` messages whatsoever
* `/run/user/1000/app/org.keepassxc.KeePassXC/org.keepassxc.KeePassXC.BrowserServer` exists
* `/run/user/1000/org.keepassxc.KeePassXC.BrowserServer` does not exist
* 
```
$ flatpak permissions webextensions                                                    
Table         Object                          App          Permissions Data
webextensions org.keepassxc.keepassxc_browser snap.firefox yes         0x00
```

**Comment by droidmonkey** [comment]
> KeepassXC window comes to front with unlocked DB, but nothing else

This shouldn't happen. Make sure you have the most up to date flatpak version of keepassxc `flatpak update` and make sure you are not pulling from a beta channel or similar. 

After that, uncheck Firefox from the browser settings, press ok, then go back in and check Firefox, press ok.

**Comment by varjolintu** [comment]
> I don't know if firefox snap on ubuntu 22.04 is currently supported or not..

I is. After you have installed the extension to Snap Firefox there should be a separate[ permissions prompt](https://149366088.v2.pressablecdn.com/wp-content/uploads/2022/07/working-as-expected.jpg) for Native Messaging.

**Comment by marcelklehr** [comment]
> flatpak update

I installed the flatpak version only yesterday, after having the same problems with the apt version of keepassxc (minus the window comes to front without anything happening) which was, however, outdated.

> After that, uncheck Firefox from the browser settings, press ok, then go back in and check Firefox, press ok.

Didn't change anything.

> there should be a separate[ permissions prompt](https://149366088.v2.pressablecdn.com/wp-content/uploads/2022/07/working-as-expected.jpg) for Native Messaging.

Wow. I've never seen that. I had the extension already installed before upgrading ubuntu to 22.04 from 20.04.

**Comment by varjolintu** [comment]
@marcelklehr Have you tried to remove the extension and install it again? Maybe the prompt is not appearing because of an update.

**Comment by marcelklehr** [comment]
> Have you tried to remove the extension and install it again?

Just tried. Doesn't help.

**Comment by dom21121** [comment]
Good morning,
I use snap firefox 110.0.1, everything was working fine until now, the connection is no longer made...
I'm on ubuntu 22.04.2, keepassxc 2.7.4 and keepassxc-browser 1.8.5.1.
...

**Comment by varjolintu** [comment]
Both KeePassXC Flatpak and Appimage are having the same problem with Firefox Snap: The browser is trying to start the `keepassxc` process inside the image instead of `keepassxc-proxy`. So maybe the Firefox Snap is starting something differently with Native Messaging and KeePassXC cannot identify if the process is trying to start the proxy or the main application.

**Comment by marcelklehr** [comment]
Is there a way to reference keepassxc-proxy inside the flatpak in the native-messaging-hosts file?

Currently it references: 
```
    "path": "/var/lib/flatpak/exports/bin/org.keepassxc.KeePassXC",
```

**Comment by droidmonkey** [comment]
- No, you need to reference the start script on order to load the proper environment

**Comment by marcelklehr** [comment]
How does keepassxc know that firefox wants to start the proxy then?

**Comment by droidmonkey** [comment]
Firefox is supposed to add the extension ID as a command line parameter

**Comment by varjolintu** [comment]
> Firefox is supposed to add the extension ID as a command line parameter

Snap Firefox isn't clearly doing that.

**Comment by vbatts** [comment]
I'm experiencing the same with KeePassXC installed via flatpak and EdgeBrowser.
(Using flatpak version, as it's newer than the v2.6.6 available on Ubuntu 22.04, which I also tried with the "Advanced" configuration to connect to `.config/microsoft-edge/NativeMessagingHosts/`, but also didn't connect)

This is a fresh install in the last week.

## Environment:

OS: Ubuntu 22.04
Edge: microsoft-edge-stable 110.0.1587.69-1
KeePassXC: org.keepassxc.KeePassXC 2.7.4 (Flatpak)
Flatpak: v1.12.7-1
KeePassXC-Browser: 1.8.5.1

## Steps

1. installed edge
2. installed keepassxc via flatpak
3. configured keepassxc to allow browser, and checked only the "edge" box
![image](https://user-images.githubusercontent.com/67049/224343388-2855583f-dc2a-4ac6-8cd7-4eac8f92ae4f.png)
4. installed keepassxc-browser extension
5. click the browser icon to "reload"
6. then error: 
![image](https://user-images.githubusercontent.com/67049/224344008-2e912523-7dae-4bcc-9f90-c5484a35c993.png)

Once I enabled Extensions "Developer Mode" and for Keepassxc-browser to "collect errors", and viewing the inspect views for "background page" I see: 
![image](https://user-images.githubusercontent.com/67049/224344623-07193fd0-cf23-4cc3-bae9-fb61d2e2295d.png)

**Comment by varjolintu** [comment]
@Leokap Check https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide#6-apparmor-with-ubuntu

**Comment by Leokap** [comment]
in #1879 I assumed you told me to install Keepass from PPA. That is what I did. I would rather like to keep my browser sandboxed... (In this guide it says what to do if Firefox is installed from PPA)

**Comment by varjolintu** [comment]
@Leokap Did the Firefox prompted you for additional permissions when you installed the extension?

**Comment by Leokap** [comment]
> I'm having this problem in snap and PPA as well. Tested each with a new Firefox Profile. At the same time plasma-browser-integration is working. About permissions:
> 
> ```
> $ flatpak permissions webextensions                                                                                                                             (base) 
> Tabla         Objeto                             Aplicación   Permisos Datos
> webextensions org.keepassxc.keepassxc_browser    snap.firefox yes      0x00
> webextensions org.kde.plasma.browser_integration snap.firefox yes      0x00
> ```

well the permission is certainly set. And it has been set by the prompt, yes.
Or are you talking about Addon Permissions within Firefox?

**Comment by varjolintu** [comment]
@Leokap So are you using KeePassXC from the PPA or Snap now? I'm a little confused. If you are using the PPA, check the troubleshooting guide again and report your findings.

**Comment by Leokap** [comment]
I have both flatpak and PPA versions of KeepassXC installed. Removed snap (and .deb=2.6.6) since I thought it might interfere. Only 2.6.6 was able to connect. I was able to get it connected with firefox snap.
I also have firefox .deb distributed in KDE Neon (probably from the firefox-PPA, don't know how to check)
the PPA-Keepass version also can't communicate with that firefox. If you want I can also test flatpak and snap of Keepass.

(I tested .deb, snap and flatpak of Firefox for Video Hardware Encoding, only snap was able to do it. (due to 12th gen intel cpu another reason why I want to stay with snap.)

**Comment by Leokap** [comment]
I'm sorry the .deb version works. I didn't see that `sudo aa-enforce /etc/apparmor.d/usr.bin.firefox` returned an error because I didn't put the text from the [guide](https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide#6-apparmor-with-ubuntu) between the parentheses.

**Comment by droidmonkey** [comment]
Narrow down to using just the flatpak keepassxc and one version of Firefox. You have a mess on your system and this isn't keepassxc at fault.

**Comment by Leokap** [comment]
you are right. I tried it on a Live USB and it worked. (firefox-snap and keepass-PPA)
I got it working on my main system by now. Don't know what exactly made the difference. I moved the files in `~/.mozilla/native-messaging-hosts/` to Trash and re-enabled the Browser integration in Keepassxc. Keepassxc wrote the exact same file into the folder (checked with `diff`) but somehow it works now.
Sorry for having wasted your time, hope this helps others with the same issue. I will definitely test my problems on a clean system first from now on.

**Comment by dom21121** [comment]
> On snap firefox 110.0.1, everything was working fine until now, the connection is no longer made... I'm on ubuntu 22.04.2, keepassxc 2.7.4 and keepassxc-browser 1.8.5.1. ...

1 - ubuntu 22.04.2 : firefox snap 110.0.01+ keepassxc 2.7.4 deb+keepassxc-browser 1.8.5.1 => does not work anymore !

2 - ubuntu 22.04.2 : firefox deb 110.0.01+ keepassxc 2.7.4 deb+keepassxc-browser 1.8.5.1 =>  everything is OK !

A try
ubuntu lunar lobster (developpement branch)  : firefox snap 110.0.01+ keepassxc 2.7.4 deb+keepassxc-browser 1.8.5.1 =>  everything is OK !

what is broken in case 1?

enough, i don't use firefox snap anymore!

**Comment by varjolintu** [comment]
@dom21121 Probably the AppArmor profile. See the previous messages for solution.

**Comment by dom21121** [comment]
ubuntu 22.10 fresh install (firefox snap + keepassxc deb), not working...

I will try to understand the AppArmor profile question.

In the initial popup that authorizes the connection (attachment), it is written:
"this permission can be changed at any time from the privacy settings"

![Capture d’écran du 2023-03-12 11-32-41](https://user-images.githubusercontent.com/121558755/224539870-a53a12f8-9b80-4799-bac8-497f5feeb594.png)


Sorry but I'm an average user and I don't see where? I would like to check.

THANKS

**Comment by droidmonkey** [comment]
That is likely done in the ubuntu store app while looking at the Firefox snap entry. Unless they have introduced a dedicated privacy settings panel for snaps.

**Comment by dom21121** [comment]
I've run
- sudo apt install flatpak
- flatpak permission-set webextensions org.keepassxc.keepassxc_browser snap.firefox yes

everything is working fine again... understand nothing... (I've keepassxc deb installed...)
Sorry to bother you

**Comment by EmPeWe** [comment]
> 1 - ubuntu 22.04.2 : firefox snap 110.0.01+ keepassxc 2.7.4 deb+keepassxc-browser 1.8.5.1 => does not work anymore !
> 
> 2 - ubuntu 22.04.2 : firefox deb 110.0.01+ keepassxc 2.7.4 deb+keepassxc-browser 1.8.5.1 => everything is OK !

It's the same setup #1 with the exact same versions mentioned. I've just dropped firefox as snap and reinstalled it as deb from PPA. Now everything is working as expected again. I wanted to get rid of snap anyway, so canonical just give me another reason.

**Comment by esemeniuc** [comment]
I'm experiencing a similar issue. Running firefox 113.0.1 on Arch with KeepassXC 2.7.5 and extension 1.8.6.1. Tried deleting `~/.mozilla/native-messaging-hosts/` and re-enabling through the UI, same issue. On the `Connected Databases` tab of the extension, pressing `Connect` doesn't do anything.

**Comment by nidomiro** [comment]
for me using the ppa version of keepassxc (instead of flatpak) solved the issue

**Comment by TCB13** [comment]
> I've run
> 
> * sudo apt install flatpak
> * flatpak permission-set webextensions org.keepassxc.keepassxc_browser snap.firefox yes
> 
> everything is working fine again... understand nothing... (I've keepassxc deb installed...) Sorry to bother you

What about ungoogled-chromium + keepassxc both installed via **flatpak**?

**Comment by varjolintu** [comment]
@TCB13 AFAIK Chromium-based Snap browsers will not work.

**Comment by TCB13** [comment]
> @TCB13 AFAIK Chromium-based browsers will not work, installed with Snap or Flatpak.

Thank you for the clarification.

It would be nice to have ungoogled-chromium working as it is one of the most private options avialable. Since it doesn't contact any servers / have updates built in I essentially would like to keep running it as a flatpak. Extra isolation is just a bonus.

**Comment by varjolintu** [comment]
For Flatpak, see: https://github.com/flatpak/xdg-desktop-portal/pull/705 and https://bugzilla.mozilla.org/show_bug.cgi?id=1621763.

**Comment by teicors** [comment]
Hi all,
in my system I have those files:
more org.keepassxc.keepassxc_browser.json_ok org.keepassxc.keepassxc_browser.json
::::::::::::::
org.keepassxc.keepassxc_browser.json_ok
::::::::::::::
```
{
    "allowed_extensions": [
        "keepassxc-browser@keepassxc.org"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "/usr/bin/keepassxc-proxy",
    "type": "stdio"
}`

Any 
```
::::::::::::::
org.keepassxc.keepassxc_browser.json
::::::::::::::
```
{
    "allowed_extensions": [
        "keepassxc-browser@keepassxc.org"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    *path": "/home/angelo/bin/KeePassXC-2.7.5-x86_64.AppImage",
    "type": "stdio"
}

```
At every restart, fie file org.keepassxc.keepassxc_browser.json is overwritten by an erroroneus string: I must replace with the "ok" file and after a restart of KeppassXC works well.
I'm working on Ubuntu 22.04, Keepass as appimage and keepassxc-browser 1.8.7 (2023-07-06).

BR,
Angelo

**Comment by varjolintu** [comment]
@teicors Your issue is not related to this thread. Please make a new one to the https://github.com/keepassxreboot/keepassxc/issues list.

**Comment by apelly** [comment]
To clarify, 

According to @varjolintu: 
> For Flatpak, see: https://bugzilla.mozilla.org/show_bug.cgi?id=1621763.

So it seems that it is fundamentally problematic for flatpacks and snaps to engage in native messaging.

If you install the [PPA](https://keepassxc.org/download/#linux) instead of the flatpack there are no such limitations and everything is rosy.

A clear note to that effect in troubleshooting tips would have saved me some hours. This appears to have been a recurring issue for so long that google can only surface incomplete, outdated, or unclear results for it.

**Comment by TCB13** [comment]
> If you install the [PPA](https://keepassxc.org/download/#linux) instead of the flatpack there are no such limitations and everything is rosy.

Will the PPA version of keepassxc with a flatpak browser such as Firefox or Ungoogled Chromium?

**Comment by varjolintu** [comment]
> > If you install the [PPA](https://keepassxc.org/download/#linux) instead of the flatpack there are no such limitations and everything is rosy.
> 
> Will the PPA version of keepassxc with a flatpak browser such as Firefox or Ungoogled Chromium?

No. It's up to Flatpak to solve problems with Native Messaging in their browser packages. For now, it seems there's no easy solutions for it.

**Comment by apelly** [comment]
> > If you install the [PPA](https://keepassxc.org/download/#linux) instead of the flatpack there are no such limitations and everything is rosy.
> 
> Will the PPA version of keepassxc with a flatpak browser such as Firefox or Ungoogled Chromium?

I have the PPA version of keepassxc, and snap firefox beta

**Comment by Scoth42** [comment]
> I've run
> 
>     * sudo apt install flatpak
> 
>     * flatpak permission-set webextensions org.keepassxc.keepassxc_browser snap.firefox yes
<snip>

This also worked for me after trying basically everything else with my snap-based Firefox including reinstalling things, resetting the database, snaps, PPAs, manual deb, and ultimately flatpak. I was about to go through the trouble of reverting back to non-snap-based Firefox (it's a pain anyway) but this did it.

So I finally have the browser integration working on Kubuntu 23.10 with default system-provided firefox snap and the PPA KeepassXC 2.7.6. I also manually installed webext-keepassxc-browser from the default system repos but I have no idea if it was necessary or not.

**Comment by vincejv** [comment]
@varjolintu was there any reason why the change from TCP sockets to native messaging support? I think years back when I was using windows, it used TCP connections. (correct me if i'm wrong? or maybe that was another keepass addon)

IMO it makes things complicated for linux users which has varying setups, the usual installation for Ubuntu new comers Firefox Snap + KeepassXC flatpak and it isn't working right out of the box

And it seems like this guide is out of date: https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide, should probably update that to reflect the events that transpired here

A lot of Chrome users will be switching the following year due to this, I hope this can be addressed at least documentation wise, on which setup is working https://arstechnica.com/gadgets/2022/09/chromes-new-ad-blocker-limiting-extension-platform-will-launch-in-2023/

**Comment by varjolintu** [comment]
> @varjolintu was there any reason why the change from TCP sockets to native messaging? I think years back when I was using windows, it used TCP connections. (correct me if i'm wrong? or maybe that was another keepass addon)
> 

In short, you cannot restrict TCP sockets to a certain user or process. Native Messaging and Unix sockets restricted to user level gives slightly better security compared to a socket that any user or process can listen with minimal effort. This was one of the biggest faults in the old KeePassHTTP: anyone could listen the TCP socket and capture keys from the encryption handshake (because it was faulty as well). If i have understood correctly, many other password managers (1Password, Bitwarden..) have later switched to Native Messaging for the same reason.

> IMO it makes things complicated for linux users which has varying setups, the usual installation for Ubuntu new comers Firefox Snap + KeepassXC appimage and it isn't working right out of the box
> 

This is due Snap's way to handle Native Messaging. I haven't looked at the AppImage case yet, but it seems Snap only knows how to start normal binaries. Anything wrapped in AppImage tries to start the main KeePassXC process instead (IIRC).
